src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,99 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
             return null;
         }
 
+        public sealed override List<(Cci.ITypeDefinition, Cci.DebugSourceDocument[])> GetTypeToDebugDocumentMap(EmitContext context)
+        {
+            var result = new List<(Cci.ITypeDefinition, Cci.DebugSourceDocument[])>();
+
+            var namespacesAndTypesToProcess = new Stack<NamespaceOrTypeSymbol>();
+            namespacesAndTypesToProcess.Push(SourceModule.GlobalNamespace);
+            while (namespacesAndTypesToProcess.Count > 0)
+            {
+                var symbol = namespacesAndTypesToProcess.Pop();
+
+                switch (symbol.Kind)
+                {
+                    case SymbolKind.Namespace:
+                        var location = GetSmallestSourceLocationOrNull(symbol);
+
+                        // filtering out synthesized symbols not having real source 
+                        // locations such as anonymous types, etc...
+                        if (location != null)
+                        {
+                            foreach (var member in symbol.GetMembers())
+                            {
+                                switch (member.Kind)
+                                {
+                                    case SymbolKind.Namespace:
+                                    case SymbolKind.NamedType:
+                                        namespacesAndTypesToProcess.Push((NamespaceOrTypeSymbol)member);
+                                        break;
+                                    default:
+                                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                                }
+                            }
+                        }
+                        break;
+                    case SymbolKind.NamedType:
+                        var typeDefinition = (Cci.ITypeDefinition)symbol.GetCciAdapter();
+                        var methodDocumentList = PooledHashSet<Cci.DebugSourceDocument>.GetInstance();
+                        GetDocumentsForMethods(methodDocumentList, typeDefinition, context);
+
+                        var debugDocuments = ArrayBuilder<Cci.DebugSourceDocument>.GetInstance();
+                        foreach (var loc in symbol.Locations)
+                        {
+                            if (!loc.IsInSource)
+                            {
+                                continue;
+                            }
+
+                            var span = loc.GetLineSpan();
+                            var debugDocument = DebugDocumentsBuilder.TryGetDebugDocument(span.Path, basePath: null);
+
+                            // We don't need to duplicate the data, if the method debug info would already contain this document
```

Comment:
>// We don't need to duplicate the data, if the method debug info would already contain this document [](http://example.com/codeflow?start=28&length=100)

not clear what "the data" means here. Better wording would say that we want to only record documents that are not already referenced by method debug info.